packages/wdio-browserstack-service/src/launcher.ts

```diff
@@ -291,20 +333,124 @@ export default class BrowserstackLauncherService implements Services.ServiceInst
+                        } else {
+                            delete (caps.capabilities as Capabilities.Capabilities)['browserstack.buildIdentifier']
+                        }
+                    } else if (capType === 'localIdentifier') {
+                        (caps.capabilities as Capabilities.Capabilities)['browserstack.localIdentifier'] = value
                     }
                 } else if (capType === 'local'){
                     (caps.capabilities as Capabilities.Capabilities)['bstack:options']!.local = true
                 } else if (capType === 'app') {
                     (caps.capabilities as Capabilities.Capabilities)['appium:app'] = value
+                } else if (capType === 'buildIdentifier') {
+                    if (value) {
+                        (caps.capabilities as Capabilities.Capabilities)['bstack:options']!.buildIdentifier = value
+                    } else {
+                        delete (caps.capabilities as Capabilities.Capabilities)['bstack:options']!.buildIdentifier
+                    }
+                } else if (capType === 'localIdentifier') {
+                    (caps.capabilities as Capabilities.Capabilities)['bstack:options']!.localIdentifier = value
                 }
             })
         } else {
             throw new SevereServiceError('Capabilities should be an object or Array!')
         }
     }
+
+    _handleBuildIdentifier(capabilities?: Capabilities.RemoteCapabilities) {
+        if (!this._buildIdentifier) {
+            this._updateCaps(capabilities, 'buildIdentifier')
+            return
+        }
+
+        if (!this._buildName || process.env.BROWSERSTACK_BUILD_NAME) {
+            if (this._buildIdentifier) {
+                this._updateCaps(capabilities, 'buildIdentifier')
+                log.warn('Skipping buildIdentifier as buildName is not passed.')
+                return
+            }
+        }
+
+        if (this._buildIdentifier && this._buildIdentifier.includes('${DATE_TIME}')){
+            const formattedDate = new Intl.DateTimeFormat('en-GB', {
+                month: 'short',
+                day: '2-digit',
+                hour: '2-digit',
+                minute: '2-digit',
+                hour12: false })
+                .format(new Date())
+                .replace(/ |, /g, '-')
+            this._buildIdentifier = this._buildIdentifier.replace('${DATE_TIME}', formattedDate)
+            this._updateCaps(capabilities, 'buildIdentifier', this._buildIdentifier)
+        }
+
+        if (!this._buildIdentifier.includes('${BUILD_NUMBER}')) return
+
+        const ciInfo = getCiInfo()
+        if (ciInfo != null) {
+            const ciBuildNumber = ciInfo.build_number
+            if (ciBuildNumber != null) {
+                this._buildIdentifier = this._buildIdentifier.replace('${BUILD_NUMBER}', 'CI '+ ciBuildNumber)
+                this._updateCaps(capabilities, 'buildIdentifier', this._buildIdentifier)
+            }
```

Comment:
```suggestion
        if (ciInfo && ciBuildNumber) {
            const ciBuildNumber = ciInfo.build_number
            this._buildIdentifier = this._buildIdentifier.replace('${BUILD_NUMBER}', 'CI '+ ciBuildNumber)
            this._updateCaps(capabilities, 'buildIdentifier', this._buildIdentifier)
```

Reply:
If we don't add null condition before `const ciBuildNumber = ciInfo.build_number` then it throws a null error. Also, the suggestion will throw uninitialised `ciBuildNumber`.

Reply:
updated suggestion